examples/erc1155/lib.rs

```diff
@@ -0,0 +1,687 @@
+            to: AccountId,
+            token_ids: Vec<TokenId>,
+            values: Vec<Balance>,
+            data: Vec<u8>,
+        ) {
+            if self.env().caller() != from {
+                assert!(
+                    self.is_approved_for_all(from, self.env().caller()),
+                    "Caller is not allowed to transfer on behalf of {:?}.",
+                    from
+                );
+            }
+
+            assert!(
+                to != AccountId::default(),
+                "Cannot send tokens to the zero-address."
+            );
+
+            assert_eq!(
+                token_ids.len(),
+                values.len(),
+                "The number of tokens being transferred does
+                 not match the number of transfer amounts."
+            );
+
+            token_ids.iter().zip(values.iter()).for_each(|(&id, &v)| {
+                self.perform_transfer(from, to, id, v, data.clone());
+            })
+        }
+
+        #[ink(message)]
+        fn balance_of(&self, owner: AccountId, token_id: TokenId) -> Balance {
+            *self.balances.get(&(owner, token_id)).unwrap_or(&0)
+        }
+
+        #[ink(message)]
+        fn balance_of_batch(
+            &self,
+            owners: Vec<AccountId>,
+            token_ids: Vec<TokenId>,
+        ) -> Vec<Balance> {
+            let mut output = Vec::new();
+            for o in &owners {
+                for t in &token_ids {
+                    let amt = self.balance_of(*o, *t);
+                    output.push(amt);
+                }
+            }
+            output
+        }
+
+        #[ink(message)]
+        fn set_approval_for_all(&mut self, operator: AccountId, approved: bool) {
+            assert!(
+                operator != self.env().caller(),
+                "An account does not need to approve themselves to transfer tokens."
+            );
+
+            if approved {
+                self.approvals.insert((self.env().caller(), operator), ());
+            } else {
+                self.approvals.remove(&(self.env().caller(), operator));
+            }
+
+            self.env().emit_event(ApprovalForAll {
+                owner: self.env().caller(),
+                operator,
+                approved,
+            });
+        }
+
+        #[ink(message)]
+        fn is_approved_for_all(&self, owner: AccountId, operator: AccountId) -> bool {
+            self.approvals.get(&(owner, operator)).is_some()
+        }
+    }
+
+    impl super::Erc1155TokenReceiver for Contract {
+        #[ink(message, selector = "0xF23A6E61")]
+        fn on_erc_1155_received(
+            &mut self,
+            _operator: AccountId,
+            _from: AccountId,
+            _token_id: TokenId,
+            _value: Balance,
+            _data: Vec<u8>,
+        ) -> Vec<u8> {
+            unimplemented!("This smart contract does not accept token transfer.")
+        }
+
+        #[ink(message, selector = "0xBC197C81")]
```

Comment:
Looking at your constant definitions above we might want to allow for identifiers here if that is possible ...

Reply:
Could be hard though since ink! proc. macros require selectors at proc. macro compile time unfortunately.